### PR TITLE
feat(warnings): add `encryptedWarning` text parameter

### PR DESCRIPTION
**Pull request by PhilipAbed**
## Changes

added `encryptedWarning` param to the GlobalConfig
if encrypted keyword is detected in config, and the `encryptedWarning` text prameter is set,
it will print the `encryptedWarning` as on log `warn` level

this will appear under the `Repository Problems` section in the dashboard

## Context

https://github.com/renovatebot/renovate/discussions/28895

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/decrypt.ts [hunk 1]
```diff
@@ -134,6 +136,13 @@ export async function decryptConfig(
   for (const [key, val] of Object.entries(config)) {
     if (key === 'encrypted' && is.object(val)) {
       logger.debug({ config: val }, 'Found encrypted config');
+
+      const encryptedWarning = GlobalConfig.get('encryptedWarning');
+      if (encryptedWarning && timesPrinted < 1) {
+        logger.warn(encryptedWarning);
+        timesPrinted++;
+      }
```
[the remark this answers is not in the document]
Comment (PhilipAbed) [reply lib/config/decrypt.ts:144]: lovely! thx

File: lib/config/options/index.ts [hunk 2]
```diff
@@ -627,6 +627,13 @@ const options: RenovateOptions[] = [
     default: true,
     globalOnly: true,
   },
+  {
+    name: 'encryptedWarning',
+    description:
+      'Display warnings about the usage of encrypted token in config',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:633]:
```suggestion
      'Renovate will print this string to the logs, and show it on your Dependency Dashboard, if it finds encrypted tokens in your Renovate configuration.',
```

File: docs/usage/self-hosted-configuration.md [hunk 3]
```diff
@@ -553,6 +553,11 @@ You can choose from the following behaviors for the `dryRun` config option:
 
 Information provided mainly in debug log level.
 
+## encryptedWarning
+
+By default, if there's an encrypted secret in the logs no warning will appear.
+By setting `encryptedWarning` text, Renovate will print the warning to the logs and will show it in the dependency dashboard.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:559]:
```suggestion
By default, Renovate does _not_ warn you if it finds an encrypted secret in the logs.

But if you set an `encryptedWarning` text, then Renovate will:

- print the `encryptedWarning` text in the logs, and
- show the `encryptedWarning` text on your Dependency Dashboard
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md [hunk 4]
```diff
@@ -553,6 +553,11 @@ You can choose from the following behaviors for the `dryRun` config option:
 
 Information provided mainly in debug log level.
 
+## encryptedWarning
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:556]: Here's the wishlish from `@rarkins`, as todos:


- [ ] A new page on `docs.renovatebot.com` explaining our preference for UI-uploaded secrets, and a section on how to migrate
- [ ] A warning message printed once per repo/job, which says something like "Support for encrypted secrets in config is deprecated by the Mend Renovate App" plus a link to this documentation page
- [ ] The warning is shown on Dependency Dashboard, and an indicator is shown in each PR mentioning that warnings were raised (not sure if we need the full message or not)

Source:

- https://github.com/renovatebot/renovate/discussions/28895#discussioncomment-9353343

Comment (PhilipAbed) [reply docs/usage/self-hosted-configuration.md:556]: 1. we would like first to support migrating secrets in production before we add it to the docs.
2. the message itself will be set on our Mend App with the doc link.
3. i will probably add the PR warnings to this PR soon, i wasn't sure if we need it or not.

File: lib/config/options/index.ts [hunk 5]
```diff
@@ -627,6 +627,13 @@ const options: RenovateOptions[] = [
     default: true,
     globalOnly: true,
   },
+  {
+    name: 'encryptedWarning',
```
[the remark this answers is not in the document]
Comment (HonkingGoose) [reply lib/config/options/index.ts:631]: Please resolve my review comment then. :wink:

File: docs/usage/self-hosted-configuration.md (hunk 4 above)
Comment (PhilipAbed) [reply docs/usage/self-hosted-configuration.md:556]: i wrote a comment in the discussion asking how to do this from product perspective, resolving this comment.

File: docs/usage/self-hosted-configuration.md [hunk 6]
```diff
@@ -553,6 +553,15 @@ You can choose from the following behaviors for the `dryRun` config option:
...
+
+By default, Renovate does _not_ warn you if it finds an encrypted secret in your config.
+But if you set an `encryptedWarning` text, then Renovate will:
+
+- print the `encryptedWarning` text in the logs
+- show the `encryptedWarning` text on your Dependency Dashboard
+- raise a warning in every pull request description
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:563]:
```suggestion
Use this if you want to stop supporting `encrypted` configuration capabilities but want to warn users first to migrate.
If set to a string value, Renovate will:

- log warnings with the `encryptedWarning` text
- show the `encryptedWarning` text on the Dependency Dashboard of any repo using `encrypted` config
- include the text as a warning in every pull request
```

File: lib/config/decrypt.ts [hunk 7]
```diff
@@ -134,6 +134,12 @@ export async function decryptConfig(
   for (const [key, val] of Object.entries(config)) {
     if (key === 'encrypted' && is.object(val)) {
       logger.debug({ config: val }, 'Found encrypted config');
+
+      const encryptedWarning = GlobalConfig.get('encryptedWarning');
+      if (is.string(encryptedWarning)) {
+        logger.once.warn(`Found encrypted config, ${encryptedWarning}`);
```
Comment (rarkins) [inline lib/config/decrypt.ts:140]:
```suggestion
        logger.once.warn(encryptedWarning);
```

Let's log the encryptedWarning text exactly as given

File: lib/config/options/index.ts [hunk 8]
```diff
@@ -627,10 +627,17 @@ const options: RenovateOptions[] = [
...
+    description:
+      'Renovate will print this string to the logs, and show it on your Dependency Dashboard, if it finds encrypted tokens in your Renovate configuration.',
```
Comment (rarkins) [inline lib/config/options/index.ts:633]:
```suggestion
      'Warning text to use if encrypted config is found.',
```

File: lib/workers/repository/errors-warnings.ts [hunk 9]
```diff
@@ -86,6 +89,52 @@ export function getDepWarningsOnboardingPR(
   return warningText;
 }
 
+export function getPrWarnings(
```
Comment (rarkins) [inline lib/workers/repository/errors-warnings.ts:92]: If this is new then it should be in a separate PR first

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/decrypt.ts (hunk 7 above)
Comment (PhilipAbed) [reply lib/config/decrypt.ts:140]: i have no indication that this message is printed when i create the PR, so i needed a static string to match in my `RepoProblems` in order to show warnings in PR.
look for `if (entry.startsWith('WARN: Found encrypted config')) {`
or should i look for the keyword `WARN:` and write that Warnings were raised regardless of this specific flag?

File: lib/workers/repository/errors-warnings.ts (hunk 9 above)
Comment (PhilipAbed) [reply lib/workers/repository/errors-warnings.ts:92]: it was the getDepPrWarning, i combined it with the encrypted warning

Comment (PhilipAbed) [reply lib/workers/repository/errors-warnings.ts:92]: separating.

File: lib/config/decrypt.ts (hunk 7 above)
Comment (PhilipAbed) [reply lib/config/decrypt.ts:140]: then Warnings will appear in PR as they appear in `repoProblems`,
ill have this in a separate PR if thats what u want

Comment (PhilipAbed) [reply lib/config/decrypt.ts:140]: if u want pr warnings to appear for any warning that goes to repoProblems there you go
https://github.com/renovatebot/renovate/pull/29205

note: this might change current behaviour for some users.

File: docs/usage/self-hosted-configuration.md (hunk 6 above)
Comment (HonkingGoose) [reply docs/usage/self-hosted-configuration.md]: Looks good to me, but can you change two things?

- Expand `repo` to `repository`
- capitalize `Pull Request`

**Comment by HonkingGoose** [comment]
I see that the blocker PR was closed in favor of discussing the needed changes.

Are we still blocked on this PR, but now on a discussion?

**Comment by PhilipAbed** [comment]
no i just fixed it, but apparently i need to do full coverage

File: docs/usage/self-hosted-configuration.md [hunk 10]
```diff
@@ -553,6 +553,16 @@ You can choose from the following behaviors for the `dryRun` config option:
...
 
+## encryptedWarning
+
+Use this if you want to stop supporting `encrypted` configuration capabilities but want to warn users first to migrate.
+
+If set to a string value, Renovate will:
+
+- log warnings with the `encryptedWarning` text
+- show the `encryptedWarning` text on the Dependency Dashboard of any repository using `encrypted` config
+- include the text as a warning in every Pull Request
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:564]:
```suggestion
If set to a string value, Renovate will log warnings with the `encryptedWarning` text, meaning the message will be visible to users such as on the Dependency Dashboard.
```

File: lib/config/options/index.ts [hunk 11]
```diff
@@ -627,10 +627,16 @@ const options: RenovateOptions[] = [
     default: true,
     globalOnly: true,
   },
+  {
+    name: 'encryptedWarning',
+    description: 'Warning text to use if encrypted config is found.',
+    type: 'string',
+    globalOnly: true,
```
Comment (rarkins) [inline lib/config/options/index.ts:634]:
```suggestion
    globalOnly: true,
    advancedUse: true,
```

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — DISMISSED** [review]

File: lib/config/options/index.ts [hunk 12]
```diff
@@ -627,10 +627,17 @@ const options: RenovateOptions[] = [
...
+    advancedUse: true,
+  },
   {
     name: 'inheritConfig',
     description:
-      'If `true`, Renovate will inherit configuration from the `inheritConfigFileName` file in `inheritConfigRepoName',
+      'If `true`, Renovate will inherit configuration from the `inheritConfigFileName` file in `inheritConfigRepoName`',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:640]:
```suggestion
      'If `true`, Renovate will inherit configuration from the `inheritConfigFileName` file in `inheritConfigRepoName`.',
```

Add missing punctuation at end of sentence.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by viceice — APPROVED** [review]
